src/Juvix/Compiler/Asm/Extra/Memory.hs

```diff
@@ -0,0 +1,184 @@
+pushTempStack :: Type -> Memory -> Memory
+pushTempStack ty = over memoryTempStack (Stack.push ty)
+
+popTempStack :: Int -> Memory -> Memory
+popTempStack n = foldr (.) id (replicate n (over memoryTempStack Stack.pop))
+
+-- | Read value stack at index `n` from the top.
+topValueStack :: Int -> Memory -> Maybe Type
+topValueStack n mem = Stack.nthFromTop n (mem ^. memoryValueStack)
+
+topValueStack' :: Int -> Memory -> Type
+topValueStack' n mem =
+  fromMaybe (error "invalid value stack access") $
+    topValueStack n mem
+
+-- | Values from top of the value stack, in order from top to bottom.
+topValuesFromValueStack :: Int -> Memory -> Maybe [Type]
+topValuesFromValueStack n mem = Stack.topValues n (mem ^. memoryValueStack)
+
+-- | Values from top of the value stack, in order from top to bottom.
+topValuesFromValueStack' :: Int -> Memory -> [Type]
+topValuesFromValueStack' n mem =
+  fromMaybe (error "invalid value stack access") $
+    Stack.topValues n (mem ^. memoryValueStack)
+
+-- | Read temporary stack at index `n` from the bottom.
+bottomTempStack :: Int -> Memory -> Maybe Type
+bottomTempStack n mem = Stack.nthFromBottom n (mem ^. memoryTempStack)
+
+bottomTempStack' :: Int -> Memory -> Type
+bottomTempStack' n mem =
+  fromMaybe (error "invalid temporary stack access") $
+    bottomTempStack n mem
+
+getArgumentType :: Offset -> Memory -> Maybe Type
+getArgumentType off mem = HashMap.lookup off (mem ^. memoryArgumentArea)
+
+getMemValueType :: InfoTable -> MemValue -> Memory -> Maybe Type
+getMemValueType tab val mem = case val of
+  DRef dr -> getDirectRefType dr mem
+  ConstrRef fld ->
+    let ci = getConstrInfo tab (fld ^. fieldTag)
+        tyargs = typeArgs (ci ^. constructorType)
+     in atMay tyargs (fld ^. fieldOffset)
+
+getDirectRefType :: DirectRef -> Memory -> Maybe Type
+getDirectRefType dr mem = case dr of
+  StackRef ->
+    topValueStack 0 mem
+  ArgRef off ->
+    getArgumentType off mem
+  TempRef off ->
+    bottomTempStack off mem
+
+getValueType' :: Member (Error AsmError) r => Maybe Location -> InfoTable -> Memory -> Value -> Sem r Type
+getValueType' loc tab mem = \case
+  ConstInt _ -> return mkTypeInteger
+  ConstBool _ -> return mkTypeBool
+  ConstString _ -> return TyString
+  ConstUnit -> return TyUnit
+  ConstVoid -> return TyUnit
+  Ref val -> case getMemValueType tab val mem of
+    Just ty -> return ty
+    Nothing -> throw $ AsmError loc "invalid memory reference"
+
+getValueType :: InfoTable -> Memory -> Value -> Maybe Type
+getValueType tab mem val =
+  case run (runError ty0) of
+    Left _ -> Nothing
+    Right ty -> Just ty
+  where
+    ty0 :: Sem '[Error AsmError] Type
+    ty0 = getValueType' Nothing tab mem val
+
+-- | Check if the value stack has at least the given height
+checkValueStackHeight' :: Member (Error AsmError) r => Maybe Location -> Int -> Memory -> Sem r ()
+checkValueStackHeight' loc n mem = do
+  unless (length (mem ^. memoryValueStack) >= n) $
+    throw $
+      AsmError
```

Comment:
How does it help modularity? To generate an error message specific to Memory.hs in the generic Error module?

Reply:
I think it helps keep the code cleaner since we only pass the necessary arguments to the error record and then in another module we format all error messages. In the scope/arity/type checkers we are doing that and it seems to be a good idea.

However, I agree that for the errors in the asm we can keep it like this since they are only for debugging